lib/components/user/notification-prefs-pane.js

```diff
@@ -28,85 +36,393 @@ const allowedNotificationChannels = [
+      // On success (status is changed to verified), update the phone number in component state and hide the code field.
+      this.setState({
+        currentNumber: phoneNumber,
+        currentNumberVerified: true,
+        isPending: false,
+        newPhoneNumber: ''
+      })
+
+      onVerified(result)
+    }
+  }
+
+  render () {
+    // All props below are Formik props (https://formik.org/docs/api/formik#props-1)
+    // except where indicated.
+    const {
+      errors,
+      phoneFormatOptions, // Provided by parent.
+      touched
+    } = this.props
+    const { currentNumber, currentNumberVerified, isEditing, isPending, newPhoneNumber } = this.state
+
+    // Here are the states we are dealing with:
+    // - First time entering a phone number/validation code (blank value, not modified)
+    //   => no color, no feedback indication.
+    // - Typing backspace all the way to erase a number/code (blank value, modified)
+    //   => red error.
+    // - Typing a phone number that doesn't match the configured phoneNumberRegEx
+    //   => red error.
+    const isPhoneInvalid = !isPossiblePhoneNumber(newPhoneNumber)
+    const showPhoneError = isPhoneInvalid && !isBlank(newPhoneNumber)
+    const phoneErrorState = showPhoneError ? 'error' : null
+    const codeErrorState = getErrorState(this.props, 'validationCode')
+
+    return (
+      <>
+        {isEditing
+          ? (
+            // FIXME: If removing the Save/Cancel buttons on the account screen,
+            // make this <FormGroup> a <form> and remove onKeyDown handler.
+            <FormGroup validationState={phoneErrorState}>
+              <ControlLabel>Enter your phone number for SMS notifications:</ControlLabel>
+              <ControlStrip>
+                <InlinePhoneInput
+                  className='form-control'
+                  country={phoneFormatOptions.countryCode}
+                  onChange={this._handleNewPhoneNumberChange}
+                  onKeyDown={this._handlePhoneNumberKeyDown}
+                  placeholder='Enter phone number'
+                  readOnly={!isEditing}
+                  type='tel'
+                  value={newPhoneNumber}
+                />
+                <Button
+                  bsStyle='primary'
+                  disabled={isPhoneInvalid}
+                  onClick={this._handleRequestCode}
+                >
+                  Send verification text
+                </Button>
+                { // Show cancel button only if a phone number is already recorded.
+                  currentNumber && <Button onClick={this._handleCancelEditNumber}>Cancel</Button>}
+                {showPhoneError && <HelpBlock>{INVALID_PHONE_MSG}</HelpBlock>}
+              </ControlStrip>
+            </FormGroup>
+          ) : (
+            <FormGroup>
+              <ControlLabel>SMS notifications will be sent to:</ControlLabel>
+              <ControlStrip>
+                <InlineStatic>
+                  {isPending
```

Comment:
If I 
1. add a phone number and click Send Verification Text
2. switch to `Don't Notify Me`
3. Switch back to SMS

The phone number and pending status disappears. I think there's an issue with `isPending`. I think its value should be calculated based on the presence of a phone number in the user object and isVerified = false. I'm guessing there are some weird things happening with state management in this component.

Reply:
Good find! Should the verification code control also be visible when the user switches back to SMS?

The Formik state is updated with the pending number in https://github.com/opentripplanner/otp-react-redux/pull/224/commits/aab453065a721588fb90219d5e9d99cbab0fb4d7.